[package] cccl/1.3: Msys2 MACHTYPE change breaks cccl's option parsing and test

### Description

Hello,

I ran into two issue when building cccl/1.3 on Windows. Both related to it's Msys2 use.

1) The test package using `self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool)` instead of `self.conf.get("tools.microsoft.bash:path", check_type=str)`
2) A fresh build of the latest msys2 conan package will have a different MACHTYPE ending in `*-cygwin` instead of `*-msys`, see the associated [MSYS2 GitHub issue](https://github.com/msys2/MSYS2-packages/issues/5232)


### Package and Environment Details

* Conan 2.8.0, 2.14.0
* Package Name/Version: **cccl/1.3**
* Operating System+version: **Windows 11**
* Build using a freshly build [msys2/ccci.latest](https://github.com/conan-io/conan-center-index/tree/5c5dac85149b205c8e12f3b41b14175b762b9a15/recipes/msys2/all/conanfile.py)
* MACHTYPE is `x86_64-pc-cygwin`



### Conan profile

[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.version=193
os=Windows

### Steps to reproduce

git clone https://github.com/conan-io/conan-center-index.git cci
cd cci/recipes/msys2/all
conan create . --version cci.latests
cd -
cd cci/recipes/cccl/all
conan create . --version 1.3 --> test breaks in two ways 1) is caught by conan 2) is a test linking error `/` used instead of `-`

### Logs

<details><summary>bash path logic error</summary>

```
======== Testing the package: Building ========
cccl/1.3 (test package): Calling build()
ERROR: cccl/1.3 (test package): Error in build() method, line 23
        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
        ConanException: [conf] tools.microsoft.bash:path must be a boolean-like object (true/false, 1/0, on/off) and value 'C:\conan_removeme\p\b\msys2778d562e9e0eb\p\bin\msys64\usr\bin\bash.exe' does not match it.
```

</details>


<details><summary>Build triplet ending in `-cygwin` resulting in options being passed incorrectly to cl</summary>

```
/out:nologo.exe
C:/conan_removeme/p/b/msys2778d562e9e0eb/p/bin/msys64/nologo
example.obj
LINK : fatal error LNK1181: cannot open input file 'C:\conan_removeme\p\b\msys2778d562e9e0eb\p\bin\msys64\nologo.obj'

ERROR: cccl/1.3 (test package): Error in build() method, line 27
        self.run(f"{cxx} {src} -o example", cwd=self.build_folder)
        ConanException: Error 2 while executing
```

</details>

--- a/recipes/cccl/all/conandata.yml
+++ b/recipes/cccl/all/conandata.yml
@@ -5,3 +5,14 @@ sources:
   "1.3":
     url: https://github.com/swig/cccl/archive/cccl-1.3.tar.gz
     sha256: 1E34E315CE3AB890D39A75FFABAACCE2E55FE5ED21591F036A45AFDA43A3E989
+patches:
+  "1.3":
+    - patch_file: "patches/0001-uname-detect-msys.patch"
+      patch_type: "official"
+      patch_description: "Use uname to detect MSYS2 intead of MATCHTYPE"
+      patch_source: "https://github.com/swig/cccl/commit/6cc59b3a14fc9c2a2301eae9809c64fbdb1b524d"
+  "1.1":
+    - patch_file: "patches/0001-uname-detect-msys.patch"
+      patch_type: "official"
+      patch_description: "Use uname to detect MSYS2 intead of MATCHTYPE"
+      patch_source: "https://github.com/swig/cccl/commit/6cc59b3a14fc9c2a2301eae9809c64fbdb1b524d"

--- a/recipes/cccl/all/conanfile.py
+++ b/recipes/cccl/all/conanfile.py
@@ -6,6 +6,9 @@ from conan.tools.microsoft import is_msvc
 import os
 
 
+required_conan_version = ">=2.1.0"
+
+
 class CcclConan(ConanFile):
     name = "cccl"
     description = "Unix cc compiler to Microsoft's cl compiler wrapper"
@@ -14,6 +17,7 @@ class CcclConan(ConanFile):
     url = "https://github.com/conan-io/conan-center-index"
     license = "GPL-3.0-or-later"
     settings = "os", "arch", "compiler", "build_type"
+    package_type = "application"
     options = {
         "muffle": [True, False],
         "verbose": [True, False],
@@ -80,13 +84,3 @@ class CcclConan(ConanFile):
         self.buildenv_info.define("CC", cccl)
         self.buildenv_info.define("CXX", cccl)
         self.buildenv_info.define("LD", cccl)
-
-        # TODO: Legacy, to be removed on Conan 2.0
-        self.env_info.PATH.append(self._cccl_dir)
-
-        self.output.info(f"Setting CC to '{cccl}'")
-        self.env_info.CC = cccl
-        self.output.info(f"Setting CXX to '{cccl}'")
-        self.env_info.CXX = cccl
-        self.output.info(f"Setting LD to '{cccl}'")
-        self.env_info.LD = cccl

--- a/recipes/cccl/all/conanfile.py
+++ b/recipes/cccl/all/conanfile.py
@@ -1,5 +1,5 @@
 from conan import ConanFile
-from conan.tools.files import get, replace_in_file, copy
+from conan.tools.files import get, replace_in_file, copy, export_conandata_patches, apply_conandata_patches
 from conan.tools.layout import basic_layout
 from conan.errors import ConanInvalidConfiguration
 from conan.tools.microsoft import is_msvc
@@ -31,6 +31,9 @@ class CcclConan(ConanFile):
     def _cccl_dir(self):
         return os.path.join(self.package_folder, "bin")
 
+    def export_sources(self):
+        export_conandata_patches(self)
+
     def layout(self):
         basic_layout(self, src_folder="src")
 
@@ -42,8 +45,8 @@ class CcclConan(ConanFile):
             raise ConanInvalidConfiguration("This recipe only supports msvc/Visual Studio.")
 
     def source(self):
-        get(self, **self.conan_data["sources"][self.version],
-                  strip_root=True, destination=self.source_folder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+        apply_conandata_patches(self)
 
     def build(self):
         cccl_path = os.path.join(self.source_folder, self.source_folder, "cccl")

--- a/recipes/cccl/all/patches/0001-uname-detect-msys.patch
+++ b/recipes/cccl/all/patches/0001-uname-detect-msys.patch
@@ -1,0 +1,42 @@
+From 6cc59b3a14fc9c2a2301eae9809c64fbdb1b524d Mon Sep 17 00:00:00 2001
+From: William S Fulton <wsf@fultondesigns.co.uk>
+Date: Fri, 7 Mar 2025 19:04:20 +0000
+Subject: [PATCH] Use uname -o to detect msys instead of x86_64-pc-msys
+
+Closes #20
+---
+ README.markdown | 2 ++
+ cccl            | 5 +++--
+ 2 files changed, 5 insertions(+), 2 deletions(-)
+
+diff --git a/README.markdown b/README.markdown
+index 1b69922..f0c3273 100644
+--- a/README.markdown
++++ b/README.markdown
+@@ -69,8 +69,10 @@ do anything but copy the cccl script, though.
+ 
+ cccl uses and hence requires ports of the following Unix utilities:
+   - bash
++  - grep
+   - sed
+   - tr
++  - uname
+ 
+ cccl is a bash shell script which makes use of these utilities.  Therefore, you
+ will need at the very least a Windows versions of these utilities.  The easiest
+diff --git a/cccl b/cccl
+index 28571fe..d681117 100755
+--- a/cccl
++++ b/cccl
+@@ -30,8 +30,9 @@ given by [OPTIONS] that Unix cc understands into parameters that cl understands.
+ EOF
+ }
+ 
+-case $MACHTYPE in
+-    *-msys)
++operating_system=$(uname -o)
++case $operating_system in
++    Msys)
+         slash="-"
+         ;;
+     *)

--- a/recipes/cccl/all/test_package/conanfile.py
+++ b/recipes/cccl/all/test_package/conanfile.py
@@ -16,18 +16,18 @@ class CcclTestConan(ConanFile):
 
     def build_requirements(self):
         self.tool_requires(self.tested_reference_str)
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
+        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             self.tool_requires("msys2/cci.latest")
 
     def build(self):
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
+        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         cxx = "cccl "
         src = os.path.join(self.source_folder, "example.cpp").replace("\\", "/")
         self.run(f"{cxx} {src} -o example", cwd=self.build_folder)
 
     def test(self):
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
+        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         if can_run(self):
             self.run("./example") #test self.run still runs in bash, so it needs "./"; seems weird but okay...

--- a/recipes/cccl/all/test_package/conanfile.py
+++ b/recipes/cccl/all/test_package/conanfile.py
@@ -11,18 +11,18 @@ class CcclTestConan(ConanFile):
 
     def build_requirements(self):
         self.tool_requires(self.tested_reference_str)
-        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", check_type=str):
             self.tool_requires("msys2/cci.latest")
 
     def build(self):
-        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         cxx = "cccl "
         src = os.path.join(self.source_folder, "example.cpp").replace("\\", "/")
         self.run(f"{cxx} {src} -o example", cwd=self.build_folder)
 
     def test(self):
-        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         if can_run(self):
             self.run("./example") #test self.run still runs in bash, so it needs "./"; seems weird but okay...

--- a/recipes/cccl/all/test_package/conanfile.py
+++ b/recipes/cccl/all/test_package/conanfile.py
@@ -9,25 +9,20 @@ class CcclTestConan(ConanFile):
     test_type = "explicit"
     win_bash = True
 
-    @property
-    def _settings_build(self):
-        # TODO: Remove for Conan v2
-        return getattr(self, "settings_build", self.settings)
-
     def build_requirements(self):
         self.tool_requires(self.tested_reference_str)
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             self.tool_requires("msys2/cci.latest")
 
     def build(self):
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         cxx = "cccl "
         src = os.path.join(self.source_folder, "example.cpp").replace("\\", "/")
         self.run(f"{cxx} {src} -o example", cwd=self.build_folder)
 
     def test(self):
-        if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         if can_run(self):
             self.run("./example") #test self.run still runs in bash, so it needs "./"; seems weird but okay...

--- a/recipes/cccl/all/test_package/conanfile.py
+++ b/recipes/cccl/all/test_package/conanfile.py
@@ -11,18 +11,18 @@ class CcclTestConan(ConanFile):
 
     def build_requirements(self):
         self.tool_requires(self.tested_reference_str)
-        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             self.tool_requires("msys2/cci.latest")
 
     def build(self):
-        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         cxx = "cccl "
         src = os.path.join(self.source_folder, "example.cpp").replace("\\", "/")
         self.run(f"{cxx} {src} -o example", cwd=self.build_folder)
 
     def test(self):
-        if self.settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
+        if not self.conf.get("tools.microsoft.bash:path", default="", check_type=str):
             return  # cccl needs a bash if there isn't a bash we can't build
         if can_run(self):
             self.run("./example") #test self.run still runs in bash, so it needs "./"; seems weird but okay...

